src/lualib/PromiseAll.ts

```diff
@@ -0,0 +1,53 @@
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all
+async function __TS__PromiseAll<T>(this: void, values: Iterable<T | PromiseLike<T>>): Promise<T[]> {
+    const results: T[] = [];
+
+    const toResolve = new LuaTable<number, PromiseLike<T>>();
+    let numToResolve = 0;
+
+    let i = 0;
+    for (const value of values) {
+        if (value instanceof __TS__Promise) {
+            if (value.state === __TS__PromiseState.Fulfilled) {
+                // If value is a resolved promise, add its value to our results array
+                results[i] = value.value;
+            } else if (value.state === __TS__PromiseState.Rejected) {
+                // If value is a rejected promise, return a rejected promise with the rejection reason
+                return Promise.reject(value.rejectionReason);
+            } else {
+                // If value is a pending promise, add it to the list of pending promises
+                numToResolve++;
+                toResolve.set(i, value);
+            }
+        } else {
+            // If value is not a promise, add it to the results array
+            results[i] = value as T;
+        }
+        i++;
+    }
+
+    // If there are no remaining pending promises, return a resolved promise with the results
+    if (numToResolve === 0) {
+        return Promise.resolve(results);
+    }
+
+    return new Promise((resolve, reject) => {
+        for (const [index, promise] of pairs(toResolve)) {
```

Comment:
Maybe replace with ipairs to preserve order? Not sure if that's mandated by the specs.

Reply:
toResolve is not an array so `ipairs` is not appropriate